libhttpserver/server.go

```diff
@@ -187,26 +189,52 @@ const portStart = 7000
 const portEnd = 8000
 const requestPathRoot = "/files/"
 
+func (s *Server) startServer() (err error) {
+	s.server = libkb.NewHTTPSrv(
+		s.g, libkb.NewPortRangeListenerSource(portStart, portEnd))
+	// Have to start this first to populate the ServeMux object.
+	if err = s.server.Start(); err != nil {
+		return err
+	}
+	s.server.Handle(requestPathRoot,
+		http.StripPrefix(requestPathRoot, http.HandlerFunc(s.serve)))
+	return nil
+}
+
+func (s *Server) monitorAppState(ctx context.Context) {
+	state := keybase1.AppState_FOREGROUND
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case state = <-s.g.AppState.NextUpdate(&state):
+			switch state {
+			case keybase1.AppState_FOREGROUND:
+				s.startServer()
+			case keybase1.AppState_BACKGROUND:
+				s.server.Stop()
+			}
+		}
+	}
+}
+
 // New creates and starts a new server.
 func New(g *libkb.GlobalContext, config libkbfs.Config) (
 	s *Server, err error) {
-	s = &Server{}
+	s = &Server{g: g, config: config}
 	s.logger = config.MakeLogger("HTTP")
 	if s.tokens, err = lru.New(tokenCacheSize); err != nil {
 		return nil, err
 	}
 	if s.fs, err = lru.New(fsCacheSize); err != nil {
 		return nil, err
 	}
-	s.config = config
-	s.server = libkb.NewHTTPSrv(
-		g, libkb.NewPortRangeListenerSource(portStart, portEnd))
-	// Have to start this first to populate the ServeMux object.
-	if err = s.server.Start(); err != nil {
+	if err = s.startServer(); err != nil {
 		return nil, err
 	}
-	s.server.Handle(requestPathRoot,
-		http.StripPrefix(requestPathRoot, http.HandlerFunc(s.serve)))
+	ctx, cancel := context.WithCancel(context.Background())
+	go s.monitorAppState(ctx)
+	s.cancel = cancel
```

Comment:
these 3 lines could happen in `s.startServer`, which would allow you not to have to re-assign `s.cancel`.